### ISSUE TITLE
Code review: security, maintainability, and error handling findings

### ISSUE THREAD

**Issue by dpalmqvist** [motsognirr/olmlx#454]
## Code review findings

A comprehensive review of the olmlx codebase across key modules (`engine/inference.py`, `engine/model_manager.py`, `cli.py`, `app.py`, `utils/streaming.py`, `routers/`, `chat/`) identified the following findings, ordered by severity.

---

### 1. [HIGH] Distributed secret exposed via SSH command line

**File:** `olmlx/cli.py:1543`

The distributed secret is injected into the SSH remote command via `shlex.quote()`, which protects against shell injection but leaves the secret visible in `ps aux` on both the coordinator and worker machines for the lifetime of the worker process:

```python
f"printf %s {shlex.quote(settings.distributed_secret)} > $SECRET_FILE",
```

Additionally, the `chmod 600` runs *after* the secret is written, so there is a window where the temp file is world-readable on the remote host.

**Recommendation:** Use SSH `SendEnv` / `AcceptEnv`, a pre-shared keyfile, or `ssh -F` with a dedicated config that sets environment variables server-side, rather than embedding the secret in the command string.

---

### 2. [HIGH] Plist writes all `OLMLX_*` environment variables including secrets

**File:** `olmlx/cli.py:1602-1618`

`_build_plist()` forwards **all** `OLMLX_*` environment variables into the launchd plist:

```python
for key, value in os.environ.items():
    if key.startswith("OLMLX_"):
        env_vars[key] = value
```

This captures `OLMLX_DISTRIBUTED_SECRET` if set. The plist is written to `~/Library/LaunchAgents/com.olmlx.plist`, which is readable by any local process. Plist storage of secrets is not a safe credential store.

**Recommendation:** Filter out known sensitive keys (`*_SECRET`, `*_KEY`, `*_TOKEN`, `*_PASSWORD`) from the plist environment variables, or require secrets to be set via a separate mechanism (keychain, config file with restricted permissions).

---

### 3. [MEDIUM] `inference.py` and `model_manager.py` are dangerously large

- **`olmlx/engine/inference.py`**: 5,277 lines
- **`olmlx/engine/model_manager.py`**: 3,864 lines

These two files together contain ~9,141 lines and encompass: lock management, inference orchestration, preflight checks, prompt caching, tokenization, chat template application, grammar setup, speculative dispatch, model loading/unloading, cache capability probing, Flash/MoE setup, TTS/Whisper loading, reranker setup, and more.

**Risk:** High cognitive load for contributors, difficult to review PRs, higher regression risk from unrelated changes touching the same file.

**Recommendation:** Split into focused modules. For `inference.py`, extract at minimum: prompt caching logic, chat template application, grammar/logits processor setup, and speculative dispatch. For `model_manager.py`, extract: speculative draft loading, Flash/MoE setup, TTS/Whisper/reranker loading into their own modules.

---

### 4. [LOW] Swallowed exception in metrics unregistration during shutdown

**File:** `olmlx/app.py:140-141`

```python
try:
    metrics_mod.REGISTRY.unregister(app.state.metrics_collector)
except Exception:
    pass
```

If the collector is already unregistered or some other failure occurs, the error is silently discarded with no log message. This could mask a real bug during test teardown or repeated `create_app()` calls.

**Recommendation:** Log at debug level before swallowing, or narrow the exception type.

---

### 5. [LOW] `BaseException` in CLI tool paths suppresses `KeyboardInterrupt`

**Files:** `olmlx/chat/builtin_tools.py:269,433`

The bash tool and search tool use `except BaseException:` to kill subprocesses before re-raising. Since these are user-facing CLI code paths (not lock-holding async context managers like the inference code), catching `BaseException` means `KeyboardInterrupt` is intercepted, the subprocess is killed, and the exception propagates — this is **functionally correct** (Ctrl+C still exits) but unnecessarily broad. A stray `SystemExit` from a background thread could also trigger the kill path.

**Recommendation:** Narrow to `except (Exception, KeyboardInterrupt):` or `except (asyncio.CancelledError, Exception):` so the intent is clear and `SystemExit` isnt intercepted.

---

### 6. [INFO] Pervasive broad `except Exception:` patterns

Across the codebase there are 41+ `except Exception:` catch-all sites (13 in `inference.py`, 11 in `model_manager.py`, 9 in `utils/streaming.py`, 6 in `cli.py`, plus dozens more). Most are in cleanup/finally paths where the intent is log-and-continue, which is reasonable. However, a few lack sufficient logging to distinguish between expected errors (e.g., a closed connection) and unexpected bugs (e.g., an `AttributeError` from a typo in a refactored field name), making debugging harder.

Notable low-logging sites:
- `olmlx/utils/streaming.py:447` — queue put failure breaks silently from the worker loop
- `olmlx/engine/model_manager.py:1200` — prompt cache eviction failure in an already-contentious TOCTOU path

**Recommendation:** In non-cleanup contexts, log the exception type and message at `debug` or `warning` level so operators can spot regressions.

---

### 7. [INFO] No test coverage for plist secret filtering or SSH secret handling

**Files:** `olmlx/cli.py:1602-1618` (plist), `olmlx/cli.py:1539-1547` (SSH secret)

The distributed tests (`tests/test_distributed.py`) test `_launch_distributed_workers` but dont assert on the secret transport mechanism. There are no tests for `_build_plist` filtering sensitive environment variables.

**Recommendation:** Add unit tests verifying that `_build_plist` filters out secret-bearing keys and that the SSH command does not include the raw secret.

---

### Overall assessment

The codebase is well-structured for a complex ML inference server. Error handling in streaming/threading paths is thorough with proper `finally` blocks, lock release, and deferred cleanup. The main areas needing attention are the two security findings (items 1 and 2 above) and the long-term maintainability risk from the oversized core modules (item 3).

**Comment by dpalmqvist** [comment]
## Triage (verified against code)

Each finding checked against the actual source. Summary:

| # | Finding | Verdict |
|---|---------|---------|
| 1 | SSH secret on command line | **REAL (core), mechanism mis-described** |
| 2 | Plist forwards all `OLMLX_*` incl. secret | **REAL** |
| 3 | inference.py / model_manager.py oversized | **REAL** (maintainability, not a bug) |
| 4 | Swallowed metrics-unregister exception | **REAL (LOW)** |
| 5 | `BaseException` in CLI tools | **NEEDS-NUANCE** |
| 6 | Broad `except` lacking logging | **Mostly mitigated — one claim wrong** |
| 7 | No tests for secret transport | **REAL** (test gap) |

### Details

**1 — NEEDS-NUANCE.** The secret *is* exposed: it lands in `remote_cmd` at `cli.py:1543` (`printf '%s' {shlex.quote(...)}`), and that whole string becomes the remote login shell's `$SHELL -c` argument — so it's visible in the **remote** host's `ps aux` (on the shell process argv, not via `printf`, which is a builtin). However, the **`chmod 600` window claim is FALSE**: line 1542 uses `SECRET_FILE=$(mktemp)`, and `mktemp` creates files mode 0600 by default — there is no world-readable window. The `chmod` is redundant belt-and-suspenders. Real mitigation = `SendEnv`/`AcceptEnv` or a keyfile to keep the secret off the argv.

**2 — REAL.** `_build_plist()` (`cli.py:1602-1606`) forwards every `OLMLX_*` env var verbatim into the launchd plist, including `OLMLX_DISTRIBUTED_SECRET` if set, in cleartext at `~/Library/LaunchAgents/com.olmlx.plist`. No secret-key filtering. (Severity tempered by single-user/localhost threat model, but behavior is exactly as described.) Fix: filter `*_SECRET`/`*_KEY`/`*_TOKEN`/`*_PASSWORD`.

**3 — REAL.** Confirmed 5,277 / 3,864 lines. Maintainability risk, not a correctness bug.

**4 — REAL (LOW).** `app.py:138-141` is a bare `except Exception: pass` with no log; shutdown-only. Add a debug log.

**5 — NEEDS-NUANCE.** `builtin_tools.py:269,433` catch `BaseException` but **re-`raise`** after killing the subprocess — so `KeyboardInterrupt` is NOT suppressed (Ctrl+C still exits). Correct cleanup-and-reraise idiom, just broader than needed. Narrowing is cosmetic.

**6 — Mostly mitigated; one cited site is wrong.** The `model_manager.py:1200` site **already logs** via `logger.debug("Concurrent close cleared prompt cache ...")` (lines 1205-1209) — the "breaks silently" framing is incorrect there. The `utils/streaming.py:447` `except Exception: break` *is* genuinely silent (real INFO-level nit).

**7 — REAL.** No assertions on the secret transport mechanism in `tests/test_distributed.py`, and no `_build_plist` filtering test.

### Recommendation
- **Worth doing:** plist secret filtering (#2); debug-log on the metrics swallow (#4) and `streaming.py:447` break (#6).
- **Consider:** `SendEnv`-based secret transport (#1); module splits (#3) as a separate refactor.
- **Correct the record:** the `chmod` window (#1) and the model_manager.py:1200 "silent" claim (#6) are not real.

**Comment by dpalmqvist** [comment]
Resolved via #457 (squash-merged to main as `a3e128e`).

- **#2 [HIGH]** plist secret filtering — credential-bearing `OLMLX_*` env vars no longer written to the cleartext launchd plist (+ test, covering **#7**).
- **#5** narrowed the two `except BaseException` cleanup sites in `chat/builtin_tools.py`.
- **#3** module splits — extracted `logits_processors.py`, `chat_templating.py`, and `speculative_loaders.py` from `inference.py` / `model_manager.py` (−18.6% / −18.8%).
- **#4** and **#6** were independently fixed on main and folded in during rebase.

Deferred (out of scope for this PR): **#1** SSH `SendEnv` secret transport, and the deeper `inference.py` prompt-cache / `model_manager` flash-MoE orchestration splits — those are coupled to the Metal-stream/lock machinery and are better as a focused follow-up.
